src/table/use-sticky-columns-context.tsx

```diff
@@ -0,0 +1,96 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+import React, { createContext, useContext, useEffect, useCallback, useRef, ReactNode } from 'react';
+
+interface StickyColumnsContextValue {
+  registerChildCallback: (callback: (entry: any) => void) => void;
+  unregisterChildCallback: (callback: (entry: any) => void) => void;
+}
+
+const StickyColumnsContext = createContext<StickyColumnsContextValue | null>(null);
+
+export const useStickyColumnsContext = () => {
+  const context = useContext(StickyColumnsContext);
+  if (!context) {
+    return { registerChildCallback: () => {}, unregisterChildCallback: () => {} };
+  }
+  const { registerChildCallback, unregisterChildCallback } = context as StickyColumnsContextValue;
+  return { registerChildCallback, unregisterChildCallback };
+};
+
+interface StickyColumnsContextProviderProps {
+  children: ReactNode;
+  wrapperRef: React.RefObject<HTMLDivElement>;
+  tableRef: React.RefObject<HTMLTableElement>;
+}
+interface StickyColumnsContextProvider {
+  (props: StickyColumnsContextProviderProps): JSX.Element;
+}
+
+export const StickyColumnsContextProvider: StickyColumnsContextProvider = ({ children, wrapperRef, tableRef }) => {
+  const childCallbacks = useRef<Set<(entry: any) => void>>(new Set());
+
+  const createScrollHandler = useCallback(() => {
```

Comment:
For people who were following this PR, but not closely, what happened to use of IntersectionObserver? Just to share learning with the team

Reply:
Good question - For this particular case (applying sticky styles when the element is stuck), the scroll event listener is actually better for the intended purposes (we're not running any heavy tasks on the callback).
The scroll event listeners reported the events more rapidly since they fire in sync with the scrolling action. The Intersection Observer runs async by nature - this means it wasn't reporting intersections at the exact moment they occur, which led to the styles being applied with just enough delay to be noticeable (the pro here being the Intersection Observer doesn't block the main thread).